### PR TITLE
Fixing MSVC build for Msbuild and Makefile generators

### PR DESCRIPTION
**Pull request by MiguelBarro**
+ Fix the generation from the current version of [protobuf](https://github.com/protocolbuffers/protobuf).
+ Included an extra workflow to generate Windows binaries from arbitrary versions on https://github.com/protobuf-c/protobuf-c/pull/688.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -117,30 +117,55 @@ jobs:
...
       - name: Build and install protobuf
-        if: steps.cache.outputs.cache-hit != 'true'
+        if: steps.protobuf-cache.outputs.cache-hit != 'true'
         run: |
           cd ~
-          C:\msys64\usr\bin\wget.exe https://github.com/protocolbuffers/protobuf/releases/download/v${{ env.PROTOBUF_VERSION }}/protobuf-cpp-${{ env.PROTOBUF_VERSION }}.zip
-          7z x protobuf-cpp-${{ env.PROTOBUF_VERSION }}.zip
-          cd ~/protobuf-${{ env.PROTOBUF_VERSION }}/cmake && mkdir build && cd build
-          cmake -G "NMake Makefiles" -DCMAKE_BUILD_TYPE=${{ matrix.build-type }} -Dprotobuf_BUILD_TESTS=OFF -DCMAKE_INSTALL_PREFIX=~/protobuf-bin -Dprotobuf_BUILD_SHARED_LIBS=${{ matrix.shared-lib }} ..
+          git clone https://github.com/protocolbuffers/protobuf.git -b v${{ env.PROTOBUF_VERSION }} protobuf
```
Comment (edmonds) [inline .github/workflows/build.yml:159]: Should use `git clone --depth=1` here which will reduce the amount of data downloaded from ~140 MB to ~6 MB.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -117,30 +117,55 @@ jobs:
...
+      - uses: actions/cache@v2
+        id: abseil-cache
+        with:
+          path: ~/abseil-bin
+          key: ${{ env.ABSEIL_VERSION }}-${{ matrix.shared-lib }}-${{ matrix.build-type}}
+      - name: Build and install abseil
+        if: steps.abseil-cache.outputs.cache-hit != 'true'
+        run: |
+          cd ~
+          git clone https://github.com/abseil/abseil-cpp.git -b ${{ env.ABSEIL_VERSION }} abseil
```
Comment (edmonds) [inline .github/workflows/build.yml:141]: `git clone --depth=1`

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -117,30 +117,55 @@ jobs:
...
+          cd ~/abseil && mkdir build && cd build
+          cmake -G "NMake Makefiles" -DCMAKE_BUILD_TYPE=${{ matrix.build-type }} -DCMAKE_INSTALL_PREFIX=~/abseil-bin -DBUILD_SHARED_LIBS=${{ matrix.shared-lib }} -DABSL_PROPAGATE_CXX_STD=ON -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded${{ matrix.build-type == 'Debug' && 'Debug' || '' }}${{ matrix.shared-lib == 'ON' && 'DLL' || '' }} -DCMAKE_CXX_STANDARD=17 ..
+          nmake
+          nmake install
+      - name: Build and install utf8 compression algorithm
+        if: matrix.shared-lib == 'OFF'
+        run: |
+          cd ~
+          git clone https://github.com/protocolbuffers/utf8_range.git utf8_range
```
Comment (edmonds) [inline .github/workflows/build.yml:150]: `git clone --depth=1`

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -90,16 +90,16 @@ jobs:
     strategy:
       matrix:
         build_type: [Debug, Release]
-        os: [macos-latest, ubuntu-20.04]
+        os: [macos-latest, ubuntu-latest]
```
Comment (edmonds) [inline .github/workflows/build.yml:93]: Let's use `ubuntu-22.04` rather than `ubuntu-latest`.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -90,16 +90,16 @@ jobs:
...
     runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v2
       - name: Install Linux dependencies
         if: startsWith(matrix.os, 'ubuntu')
-        run: sudo apt-get install -y protobuf-compiler libprotobuf-dev libprotoc-dev
+        run: sudo apt-get install -y protobuf-compiler libprotobuf-dev libprotoc-dev libabsl-dev
```
Comment (edmonds) [inline .github/workflows/build.yml:99]: Why do we need `libabsl-dev` on Ubuntu? The absolute newest version of protobuf in the Debian/Ubuntu ecosystem is currently protobuf 3.21.12 in Debian unstable which pre-dates the absl dependency in newer versions of protobuf.

**Comment by edmonds** [comment]
Hi, @MiguelBarro!

Thanks for this, it looks like this makes the MSVC CMake builds succeed. I left a few comments inline.

> • Include an extra workflow to generate Windows binaries from arbitrary versions.

I think shipping MSVC artifacts is separate from fixing the build on MSVC, and I'm not entirely sure if we should ship binary artifacts at all. Could you split this change out (the `.github/workflows/msvc_bin.yml` file) so that it can be submitted and reviewed as a separate PR?

Thanks!

File: .github/workflows/build.yml (hunk 1 above)
Comment (MiguelBarro) [reply .github/workflows/build.yml]: agreed 👍

File: .github/workflows/build.yml (hunk 2 above)
Comment (MiguelBarro) [reply .github/workflows/build.yml]: agreed 👍

File: .github/workflows/build.yml (hunk 3 above)
Comment (MiguelBarro) [reply .github/workflows/build.yml]: agreed 👍

File: .github/workflows/build.yml (hunk 4 above)
Comment (MiguelBarro) [reply .github/workflows/build.yml]: agreed 👍

File: .github/workflows/build.yml (hunk 5 above)
Comment (MiguelBarro) [reply .github/workflows/build.yml]: That's right ... removed

**Comment by MiguelBarro** [comment]
I will create another pull for the binary generator

**Review by edmonds — APPROVED** [review]

**Pull request merged**